v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/transform/ChangeEventTransformerDoFn.java

```diff
@@ -0,0 +1,223 @@
+      Metrics.counter(ChangeEventTransformerDoFn.class, "Filtered events");
+
+  private final Counter transformedEvents =
+      Metrics.counter(ChangeEventTransformerDoFn.class, "Transformed events");
+
+  private final Counter skippedEvents =
+      Metrics.counter(ChangeEventTransformerDoFn.class, "Skipped events");
+  private final Counter failedEvents =
+      Metrics.counter(ChangeEventTransformerDoFn.class, "Other permanent errors");
+
+  private final Counter customTransformationException =
+      Metrics.counter(ChangeEventTransformerDoFn.class, "Custom Transformation Exceptions");
+
+  private final Distribution applyCustomTransformationResponseTimeMetric =
+      Metrics.distribution(
+          ChangeEventTransformerDoFn.class, "apply_custom_transformation_impl_latency_ms");
+
+  public static ChangeEventTransformerDoFn create(
+      Schema schema,
+      TransformationContext transformationContext,
+      String sourceType,
+      CustomTransformation customTransformation) {
+    return new AutoValue_ChangeEventTransformerDoFn(
+        schema, transformationContext, sourceType, customTransformation);
+  }
+
+  /** Setup function connects to Cloud Spanner. */
+  @Setup
+  public void setup() {
+    mapper = new ObjectMapper();
+    mapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
+    datastreamToSpannerTransformer =
+        CustomTransformationImplFetcher.getCustomTransformationLogicImpl(customTransformation());
+    changeEventSessionConvertor =
+        new ChangeEventSessionConvertor(schema(), transformationContext(), sourceType());
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) {
+    FailsafeElement<String, String> msg = c.element();
+    processedEvents.inc();
+
+    try {
+
+      JsonNode changeEvent = mapper.readTree(msg.getPayload());
+
+      if (!schema().isEmpty()) {
+        schema().verifyTableInSession(changeEvent.get(EVENT_TABLE_NAME_KEY).asText());
+        changeEvent = changeEventSessionConvertor.transformChangeEventViaSessionFile(changeEvent);
+      }
+
+      // If custom jar is specified apply custom transformation to the change event
+      if (datastreamToSpannerTransformer != null) {
+        MigrationTransformationResponse migrationTransformationResponse =
+            getCustomTransformationResponse(changeEvent);
+        if (migrationTransformationResponse.isEventFiltered()) {
+          filteredEvents.inc();
+          c.output(DatastreamToSpannerConstants.FILTERED_EVENT_TAG, msg.getPayload());
+          return;
+        }
+        changeEvent =
+            ChangeEventToMapConvertor.transformChangeEventViaCustomTransformation(
+                changeEvent, migrationTransformationResponse.getResponseRow());
+      }
+      transformedEvents.inc();
+      c.output(
+          DatastreamToSpannerConstants.TRANSFORMED_EVENT_TAG,
+          FailsafeElement.of(changeEvent.toString(), changeEvent.toString()));
+    } catch (DroppedTableException e) {
```

Comment:
Can we write some UTs to cover the exception scenarios?

Reply:
+1

Reply:
added